Spike on computing an edge-restricted transitive closure in Postgres

tuzz opened the pull request:
I've opened a pull request so that we can review the approach. This code shouldn't be merged.


```ruby
require 'benchmark';

hash = LinkSet.all.offset(200000).each.with_object({}) do |link_set, hash|
  results = nil;
  time_taken = Benchmark.realtime do
    results = Queries::GetDependents.call(link_set.content_id)
  end
  next if results.size.zero?
  hash[link_set.content_id] = [time_taken, results.size]
  puts "#{link_set.content_id} : #{results.size} results : query took #{time_taken.round(4)} seconds"
end; nil
```

danielroseman commented:
Excellent stuff. The docs on WITH RECURSIVE mention that these queries can loop indefinitely if the data has cycles. Do we need to put in a guard for that?

tuzz commented:
Good question – from the docs:

> When working with recursive queries it is important to be sure that the recursive part of the query will eventually return no tuples, or else the query will loop indefinitely. Sometimes, using UNION instead of UNION ALL can accomplish this by discarding rows that duplicate previous output rows. However, often a cycle does not involve output rows that are completely duplicate: it may be necessary to check just one or a few fields to see if the same point has been reached before.

In this case, output rows are completely duplicate of rows that have already been retrieved, so it will terminate.

boffbowsh commented:
Added a quick EXPLAIN for this: http://explain.depesz.com/s/Nyd

It looks ok, but there might be some easy optimisations.

tuzz commented:
Thanks @boffbowsh – I've reworked it a bit and added indexes based on looking through the explain.

Here's how it performs for a query restricted to "parent" link type:

```
453affe4-5ebb-4f44-9f0a-44cfd32dc934 : 1 results : query took 0.0008 seconds
17e4ab26-ee1f-4383-a345-d165c0b75fbf : 1 results : query took 0.0008 seconds
cf8737a8-45b0-4a9d-a3df-3fef1bdbca59 : 1 results : query took 0.0009 seconds
4b5e2787-7429-4547-9550-123ec1dade70 : 1 results : query took 0.0017 seconds
7bd8a095-2a4c-4fed-ad1d-f12c85356906 : 1 results : query took 0.0015 seconds
0146d0ed-dc6b-490f-a0e0-2edb8927a134 : 1 results : query took 0.0015 seconds
b4b132aa-93af-4604-9dae-b6de2d141427 : 1 results : query took 0.0007 seconds
08e5731b-5286-49d3-a5bd-e859862b91c8 : 1 results : query took 0.0014 seconds
c0d6920f-ca3b-4eaf-81c8-705d3b483e4e : 1 results : query took 0.0008 seconds
fe8a3ab6-d0bd-4050-966f-488dec46fd49 : 1 results : query took 0.0011 seconds
6edb43ee-3325-4c12-9669-93b7ce015c48 : 1 results : query took 0.0008 seconds
3e18dd40-856d-49aa-9bce-3d40906a2240 : 1 results : query took 0.0011 seconds
c2406423-04c0-4e6e-a6c6-847a1287cb59 : 1 results : query took 0.001 seconds
d677da19-6e25-4790-b25f-035e344980df : 1 results : query took 0.0007 seconds
93018f41-1b37-4cb4-b90d-be8225dd8fac : 1 results : query took 0.0009 seconds
9c0a5d6f-a0d5-4c2a-9bbd-d8e5cd3b323e : 1 results : query took 0.0007 seconds
79918ebf-6d58-437b-adc4-93d9328167bf : 1 results : query took 0.0007 seconds
38a18dbb-7fe1-4e2e-8587-8627610ba901 : 1 results : query took 0.0012 seconds
7f12a3bd-510c-4dfb-957d-ea30e51cc392 : 1 results : query took 0.0008 seconds
dc6d2e0e-8f5d-4c3f-aaea-c890e07d0cf8 : 1 results : query took 0.0005 seconds
66650627-a8f6-47b1-a06e-ac55700f2e57 : 1 results : query took 0.0007 seconds
3c3d7d5a-91cf-44df-a8df-cb98a601e44e : 1 results : query took 0.0012 seconds
3503917f-cf1b-45a2-b5ee-60ee32e66d82 : 1 results : query took 0.0008 seconds
bcf4b464-08b7-41e9-88c5-2c990be6961c : 1 results : query took 0.0007 seconds
d55fc267-db46-4ab5-be2c-1dc49e811dc3 : 8 results : query took 0.0008 seconds
9faf140c-e6bb-4b79-a83e-ce69672b74fd : 7 results : query took 0.0008 seconds
502eac33-76ad-499e-b15f-a1544be548c5 : 10 results : query took 0.0009 seconds
e184c8d6-7142-4637-ae1e-3270247e0f4f : 1 results : query took 0.0008 seconds
7dea08e2-13a1-4ba7-a1ef-bc8e7fb218dd : 3 results : query took 0.0011 seconds
d4e23b21-9b47-4847-b95a-c9fc95e5f22f : 8 results : query took 0.0008 seconds
14887b49-b1ed-4c95-9daa-031c93a29721 : 2 results : query took 0.0007 seconds
47f86db5-4641-4aca-b3fc-c15fcfa4da46 : 1 results : query took 0.0012 seconds
c3037eb5-b3a0-4585-b568-f6e7a59ef7d5 : 4 results : query took 0.0007 seconds
```

Here's how it performs for an unrestricted query (any link_type):

```
0f8c5b0f-5660-4c1b-8f8a-4d33787cc59e : 15 results : query took 0.0014 seconds
c9976d87-0d3b-4349-9dec-fbb009871269 : 497 results : query took 0.0069 seconds
78d0cadd-ce01-4c74-8f7d-2a0fb9364c7c : 2 results : query took 0.0014 seconds
a90d4fac-5656-4745-9c16-08542f42c3ac : 66 results : query took 0.0019 seconds
020cdae2-ee93-4fca-bfb7-3f5f8f5751d1 : 228 results : query took 0.0039 seconds
c75939ac-59b4-4a00-ad23-107d6cecf95b : 9 results : query took 0.0014 seconds
79918ebf-6d58-437b-adc4-93d9328167bf : 1 results : query took 0.001 seconds
38a18dbb-7fe1-4e2e-8587-8627610ba901 : 1 results : query took 0.0013 seconds
7f12a3bd-510c-4dfb-957d-ea30e51cc392 : 1 results : query took 0.0007 seconds
dc6d2e0e-8f5d-4c3f-aaea-c890e07d0cf8 : 32 results : query took 0.0012 seconds
66650627-a8f6-47b1-a06e-ac55700f2e57 : 1 results : query took 0.0009 seconds
3c3d7d5a-91cf-44df-a8df-cb98a601e44e : 1 results : query took 0.0007 seconds
3503917f-cf1b-45a2-b5ee-60ee32e66d82 : 1 results : query took 0.0007 seconds
445370d5-b01d-40a9-a660-792365f29362 : 2 results : query took 0.0007 seconds
c03485d4-e94d-400d-b1b2-6fd7b6c739a0 : 32 results : query took 0.0011 seconds
bcf4b464-08b7-41e9-88c5-2c990be6961c : 1 results : query took 0.0007 seconds
f4c3a0be-7a30-11e4-a3cb-005056011aef : 1 results : query took 0.0008 seconds
d55fc267-db46-4ab5-be2c-1dc49e811dc3 : 385 results : query took 0.0056 seconds
0d76003d-76ba-4af5-a19e-31ab3f8ba689 : 21 results : query took 0.0019 seconds
e6b3648e-0b2d-4b49-864a-2079bf334ec0 : 7 results : query took 0.0013 seconds
efc80542-f10d-42cf-b798-8bbf87327384 : 21 results : query took 0.0014 seconds
8c323bdf-0660-4155-9d31-6c6bf9322b00 : 10 results : query took 0.0011 seconds
948fc362-e8bc-4127-b870-294e61c2a659 : 10 results : query took 0.0015 seconds
20c5a5bd-fcd0-427a-866f-79d2a788f6a7 : 10 results : query took 0.0013 seconds
79a23304-dbe9-44d6-b6da-0827b4fa8bdf : 9 results : query took 0.001 seconds
e049df4f-68bf-4ca2-89b8-e4ee26af765e : 9 results : query took 0.0016 seconds
28d7a893-a98d-470a-b7b6-21d425be4317 : 1 results : query took 0.0006 seconds
77589ce6-f1af-49de-9449-ef3faa9b9409 : 8 results : query took 0.0013 seconds
```

It's worth noting that the links probably aren't set up the way they will be when we have large breadcrumb ancestries. Will explore this in more detail.

There's probably more we can do to tune the query, too.

boffbowsh commented on app/queries/get_dependents.rb:
```diff
@@ -0,0 +1,44 @@
+      ).to_a
+    end
+
+    # Uses Postgres's WITH RECURSIVE feature to compute the transitive closure
```
Can you please explain "transitive closure" in plainer English? This is a complicated query below and I (and future others) won't fully understand what it's doing and how it's doing it.

boffbowsh commented:
Looks better :clap:

tuzz commented:
I've tried to make the comment a bit clearer @boffbowsh. Let me know if it makes sense! :-)

tuzz commented:
I had a look at getting the depth of the dependency and also the path to the dependency through the graph. This works, but it slows the query down by about 50%:

```sql
          WITH RECURSIVE dependents(content_id, link_type, path, depth, cycle) AS (
            SELECT
              '#{content_id}'::TEXT, '#{link_type}', ARRAY['#{content_id}'], 0, false
          UNION
              SELECT DISTINCT
                link_sets.content_id,
                dependents.link_type,
                path || link_sets.content_id::TEXT,
                dependents.depth + 1,
                link_sets.content_id = ANY(path)
              FROM dependents
              JOIN links
                ON links.target_content_id = dependents.content_id
            #{'AND links.link_type = dependents.link_type' if link_type}
              JOIN link_sets
                ON link_sets.id = links.link_set_id
              WHERE NOT cycle
          )
          SELECT
            content_id
          FROM dependents
          WHERE content_id != '#{content_id}'
```

boffbowsh commented:
Calculating the depths is probably most useful when the content items have already been resolved and individual items fetched. Then we can just walk those items to calculate depths as needed and only for breadcrumbs.

boffbowsh commented:
Also :+1: for the comment clarification, thanks for that.

tuzz commented:
Please don't delete this branch as we will refer to it on an upcoming ticket:
https://trello.com/c/CjQNKShl/99-4-calculate-the-set-of-content-items-that-need-to-be-updated

Closing for now.

tuzz commented:
For reference, the approach for computing the depth was based on:
https://www.postgresql.org/docs/8.4/static/queries-with.html
